File: ios/Sources/TaigiKeyboard/NextWord/NextWordEngine.swift
import Foundation

// MARK: - Shared-Core Candidate

// Pure logic, Foundation-only. Eligible for cross-platform extraction.

/// Pure decision + filtering layer for NextWord prediction.
///
/// Split from `NextWordController` (G5-impl): the controller remains the
/// iOS platform executor (Timer, @MainActor, settings read, generation
/// bookkeeping, UI fan-out); this engine owns validation, state
/// transitions, association-window math, compound-word splitting, and
/// prediction filtering.
///
/// **Invariants** (see `docs/architecture/nextword-engine-boundary.md`):
/// - No clock reads. Callers supply `nowMs` via `NextWordDecisionInput`.
/// - No settings provider. Callers pass a `NextWordEngineSettings` value.
/// - No Timer / DispatchQueue / Task. Scheduling is described as effects
///   the executor interprets.
/// - Any state-invalidating intent bumps `currentGeneration` so late
///   async prediction results can be discarded by the executor.
enum NextWordEngine {
    // MARK: - Constants

    /// Strict-less-than association window. A gap of exactly 10_000 ms does
    /// NOT record; negative deltas (clock skew) also do not record.
    static let associationTimeoutMs: Int64 = 10000

    /// Context timeout scheduled via `.rescheduleContextTimeout(after:)`.
    static let contextTimeoutSeconds: TimeInterval = 30.0

    /// Punctuation that ends a sentence and resets NextWord state.
    static let sentenceEndPunctuation: Set<Character> = ["。", "！", "？", ".", "!", "?"]

    /// Superset of sentence-end: any char in this set aborts `wordSelected`
    /// without recording or predicting (punctuation/whitespace/number paths).
    private static let noisePunctuation: Set<Character> = Set(
        "。！？.!?，,、；;：:「」『』\"\"\u{2018}\u{2019}（）()【】[]{}—–-～~…·",
    )

    // MARK: - Entry Points

    /// Decide what the platform executor should do in response to a single
    /// lifecycle event. Pure function: same inputs → same outputs.
    static func decide(
        intent: NextWordIntent,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ) -> NextWordOutcome {
        switch intent {
        case let .wordSelected(text, roman, requireRomanMode, triggerPrediction):
            decideWordSelected(
                text: text,
                roman: roman,
                requireRomanMode: requireRomanMode,
                triggerPrediction: triggerPrediction,
                state: state,
                input: input,
            )
        case let .backspace(lastChar):
            decideBackspace(lastChar: lastChar, state: state, input: input)
        case .contextTimeoutFired:
            decideContextTimeoutFired(state: state)
        case .clearForNewComposing:
            decideClearForNewComposing(state: state)
        case .resetFull:
            decideResetFull(state: state)
        }
    }

    /// Filter + shape raw service predictions into UI-ready values.
    /// Pure function mirroring today's `makePredictions(from:)` but typed
    /// on the shared DTO so it can move to shared-core without depending
    /// on `NextWordService`.
    static func filterPredictions(
        _ raw: [RawNextWordPrediction],
        settings: NextWordEngineSettings,
    ) -> [EnginePrediction] {
        raw.compactMap { prediction in
            if !settings.isTranslateSwapped, prediction.tl.isEmpty {
                return nil
            }

            let roman = settings.inputMode == .poj
                ? RomanizationConverter.tlToPOJ(prediction.tl)
                : prediction.tl
            let text = roman.isEmpty ? prediction.hanzi : roman
            let subtitle: String? = roman.isEmpty ? nil : prediction.hanzi

            return EnginePrediction(
                text: text,
                subtitle: subtitle,
                hanzi: prediction.hanzi,
                tl: prediction.tl,
            )
        }
    }

    // MARK: - Pure Helpers (exposed for testing)

    /// Association window check. Strict `<` at 10s; negative deltas (clock
    /// skew / wrapping) return false — a deliberate fix vs the pre-split
    /// code, which would have returned true for a future-relative state.
    static func shouldRecordAssociation(
        state: NextWordPersistedState,
        nowMs: Int64,
    ) -> Bool {
        guard state.lastSelectedWord != nil else { return false }
        let delta = nowMs - state.lastSelectionTimeMs
        guard delta >= 0 else { return false }
        return delta < associationTimeoutMs
    }

    /// Split a compound word on `-` (e.g. "tshit-niû" → ["tshit", "niû"]).
    /// Empty parts are dropped so `"-a-"` yields `["a"]`, not `["", "a", ""]`.
    static func splitCompound(_ word: String) -> [String] {
        guard !word.isEmpty else { return [] }
        return word.split(separator: "-").map(String.init).filter { !$0.isEmpty }
    }

    /// Build sequential bigram pairs from a compound word. Order preserved
    /// so the executor can feed them to `NextWordService.recordAssociation`
    /// sequentially (parallel writes would race on the SQLite UNIQUE index).
    static func compoundAssociationPairs(
        displayText: String,
        roman: String,
    ) -> [NextWordAssociationPair] {
        let parts = splitCompound(displayText)
        let romanParts = splitCompound(roman)
        guard parts.count > 1 else { return [] }

        var pairs: [NextWordAssociationPair] = []
        pairs.reserveCapacity(parts.count - 1)
        for i in 0 ..< (parts.count - 1) {
            let prevTl = romanParts.indices.contains(i) ? romanParts[i] : ""
            let nextTl = romanParts.indices.contains(i + 1) ? romanParts[i + 1] : ""
            pairs.append(NextWordAssociationPair(
                prev: parts[i],
                prevTl: prevTl,
                next: parts[i + 1],
                nextTl: nextTl,
            ))
        }
        return pairs
    }

    // MARK: - Intent Handlers

    private static func decideWordSelected(
        text: String,
        roman: String,
        requireRomanMode: Bool,
        triggerPrediction: Bool,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ) -> NextWordOutcome {
        // Enter commits raw romanization only; skip entirely in Hanji mode.
        if requireRomanMode, input.settings.isTranslateSwapped {
            return NextWordOutcome(newState: state, effects: [])
        }

        // Noise text (punctuation / whitespace / digits) never records or
        // predicts. Sentence-end punctuation is a *subset* of noise: it
        // additionally resets state + clears UI, so we branch on it inside
        // the noise guard before returning.
        guard !text.isEmpty, !isNoiseText(text) else {
            if isSentenceEndPunctuation(text) {
                return resetAndClearPredictions(state: state)
            }
            return NextWordOutcome(newState: state, effects: [])
        }

        // pojToTL is idempotent on TL input — safe for POJ and TPS alike.
        let textTl = RomanizationConverter.pojToTL(roman)
        let prevTl = RomanizationConverter.pojToTL(state.lastSelectedRoman ?? "")

        var effects: [NextWordOutcome.Effect] = []

        if input.settings.isAssociationRecordingEnabled {
            if shouldRecordAssociation(state: state, nowMs: input.nowMs),
               let prevWord = state.lastSelectedWord
            {
                effects.append(.recordAssociation(NextWordAssociationPair(
                    prev: prevWord,
                    prevTl: prevTl,
                    next: text,
                    nextTl: textTl,
                )))
            }
            let compound = compoundAssociationPairs(displayText: text, roman: textTl)
            if !compound.isEmpty {
                effects.append(.recordCompoundAssociations(compound))
            }
        }

        effects.append(.rescheduleContextTimeout(after: contextTimeoutSeconds))

        var newState = state
        newState.lastSelectedWord = text
        newState.lastSelectedRoman = textTl
        newState.lastSelectionTimeMs = input.nowMs
        newState.currentGeneration &+= 1

        if triggerPrediction {
            effects.append(.queryPredictions(
                word: text,
                roman: textTl,
                generation: newState.currentGeneration,
            ))
        }

        return NextWordOutcome(newState: newState, effects: effects)
    }

    private static func decideBackspace(
        lastChar: String,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ) -> NextWordOutcome {
        var newState = state
        newState.lastSelectedWord = lastChar
        newState.lastSelectedRoman = nil
        newState.lastSelectionTimeMs = input.nowMs
        newState.currentGeneration &+= 1

        return NextWordOutcome(
            newState: newState,
            effects: [
                .queryPredictions(word: lastChar, roman: "", generation: newState.currentGeneration),
            ],
        )
    }

    private static func decideContextTimeoutFired(
        state: NextWordPersistedState,
    ) -> NextWordOutcome {
        resetAndClearPredictions(state: state)
    }

    private static func decideClearForNewComposing(
        state: NextWordPersistedState,
    ) -> NextWordOutcome {
        var newState = state
        newState.isShowing = false
        newState.currentGeneration &+= 1

        let effects: [NextWordOutcome.Effect] = state.isShowing
            ? [.clearPredictionsUI(generation: newState.currentGeneration)]
            : []
        return NextWordOutcome(newState: newState, effects: effects)
    }

    private static func decideResetFull(
        state: NextWordPersistedState,
    ) -> NextWordOutcome {
        resetAndClearPredictions(state: state)
    }

    /// Shared reset path used by sentence-end punctuation, context timeout,
    /// and `resetFull` intents. Cancels the Timer, optionally clears the
    /// UI, zeroes state, bumps generation.
    private static func resetAndClearPredictions(
        state: NextWordPersistedState,
    ) -> NextWordOutcome {
        var newState = NextWordPersistedState.initial
        newState.currentGeneration = state.currentGeneration &+ 1

        var effects: [NextWordOutcome.Effect] = [.cancelContextTimeout]
        if state.isShowing {
            effects.append(.clearPredictionsUI(generation: newState.currentGeneration))
        }
        return NextWordOutcome(newState: newState, effects: effects)
    }

    // MARK: - Classification

    /// Punctuation / whitespace / pure-digit text never triggers NextWord.
    static func isNoiseText(_ text: String) -> Bool {
        guard let firstChar = text.first else { return true }
        if noisePunctuation.contains(firstChar) { return true }
        if firstChar.isWhitespace { return true }
        if text.allSatisfy({ $0.isASCII && $0.isNumber }) { return true }
        return false
    }

    static func isSentenceEndPunctuation(_ text: String) -> Bool {
        guard let firstChar = text.first else { return false }
        return sentenceEndPunctuation.contains(firstChar)
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/core/nextword/NextWordEngine.kt
package com.siansiansu.taigikeyboard.ime.core.nextword

import com.siansiansu.taigikeyboard.ime.dictionary.TaigiPhonetics

// region Shared-Core Candidate
// Pure logic, Kotlin stdlib only. Eligible for cross-platform extraction.
// endregion

/**
 * Pure decision + filtering layer for NextWord prediction.
 *
 * Mirrors iOS `NextWord/NextWordEngine.swift`. The executor
 * ([com.siansiansu.taigikeyboard.ime.text.smartbar.NextWordHandler]) owns
 * the coroutine-scheduled context-timeout job, `EngineSettingsProvider`
 * reads, and UI callbacks; this engine owns validation, state transitions,
 * association-window math, compound-word splitting, and prediction
 * filtering.
 *
 * Invariants (see `docs/architecture/nextword-engine-boundary.md`):
 * - No clock reads. Callers supply `nowMs` via [NextWordDecisionInput].
 * - No settings provider. Callers pass a [NextWordEngineSettings] value.
 * - No coroutines / Dispatchers / Job. Scheduling is described as effects
 *   the executor interprets.
 * - Any state-invalidating intent bumps [NextWordPersistedState.currentGeneration]
 *   so late async prediction results can be discarded by the executor.
 */
object NextWordEngine {
    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/NextWord/NextWordEngine.swift:27.
    // Drift causes silent divergence. A8-sweep tightens the marker wording across constants.
    /** Strict-less-than association window. 10_000 ms gap does NOT record; negative deltas also skipped. */
    const val ASSOCIATION_TIMEOUT_MS: Long = 10_000L

    // CROSS-PLATFORM INVARIANT — mirrors ios/Sources/TaigiKeyboard/NextWord/NextWordEngine.swift:30.
    // Drift causes silent divergence. Android uses milliseconds (Kotlin `delay(Long)` idiom);
    // iOS uses `TimeInterval` seconds. Same 30-second wait.

    /** Context-timeout window. Scheduled via `Effect.RescheduleContextTimeout(afterMs)`. */
    const val CONTEXT_TIMEOUT_MS: Long = 30_000L

    /** Punctuation that ends a sentence and resets NextWord state. */
    val sentenceEndPunctuation: Set<Char> =
        setOf('。', '！', '？', '.', '!', '?')

    /**
     * Superset of sentence-end: any char in this set aborts `WordSelected`
     * without recording or predicting. Android preserves its pre-A5-impl
     * superset (includes Chinese punctuation, quote variants, digits via
     * [isNoiseText]'s all-ASCII-digits check). iOS limits the set to the
     * punctuation characters only. Intentional divergence — documented in
     * `nextword-engine-boundary.md` §13 Android addendum.
     */
    private val noisePunctuation: Set<Char> =
        setOf(
            '。',
            '！',
            '？',
            '.',
            '!',
            '?',
            '，',
            ',',
            '、',
            '；',
            ';',
            '：',
            ':',
            '「',
            '」',
            '『',
            '』',
            '"',
            '"',
            '\'',
            '（',
            '）',
            '(',
            ')',
            '【',
            '】',
            '[',
            ']',
            '{',
            '}',
            '—',
            '–',
            '-',
            '～',
            '~',
            '…',
            '·',
            ' ',
            '　',
        )

    /**
     * Decide what the platform executor should do in response to a single
     * lifecycle event. Pure function: same inputs → same outputs.
     */
    fun decide(
        intent: NextWordIntent,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ): NextWordOutcome =
        when (intent) {
            is NextWordIntent.WordSelected -> {
                decideWordSelected(
                    text = intent.text,
                    roman = intent.roman,
                    requireRomanMode = intent.requireRomanMode,
                    triggerPrediction = intent.triggerPrediction,
                    state = state,
                    input = input,
                )
            }

            is NextWordIntent.Backspace -> {
                decideBackspace(lastChar = intent.lastChar, state = state, input = input)
            }

            NextWordIntent.ContextTimeoutFired -> {
                resetAndClearPredictions(state)
            }

            NextWordIntent.ClearForNewComposing -> {
                decideClearForNewComposing(state)
            }

            NextWordIntent.ResetFull -> {
                resetAndClearPredictions(state)
            }
        }

    /**
     * Filter + shape raw service predictions into UI-ready values. Pure
     * function mirroring today's wrapper-side `updateCandidatesWithPredictions`
     * logic but typed on the shared DTO so it can move to shared-core
     * without depending on platform types.
     *
     * Divergence from iOS filter: Android checks `inputMode == "tl"` to
     * decide whether to convert roman; iOS checks `inputMode == .poj`. Same
     * outcome (convert when mode is POJ), different dispatch. Kept
     * consistent with pre-A5 Android behavior.
     */
    fun filterPredictions(
        raw: List<RawNextWordPrediction>,
        settings: NextWordEngineSettings,
    ): List<EnginePrediction> {
        val useTl = settings.inputMode == "tl"
        return raw.mapNotNull { prediction ->
            val roman =
                if (useTl) {
                    prediction.tl
                } else {
                    TaigiPhonetics.tlDisplayToPOJDisplay(prediction.tl)
                }

            if (!settings.isTranslateSwapped && roman.isEmpty()) return@mapNotNull null

            val text = if (roman.isEmpty()) prediction.hanzi else roman
            val subtitle: String? = if (roman.isEmpty()) null else prediction.hanzi

            EnginePrediction(
                text = text,
                subtitle = subtitle,
                hanzi = prediction.hanzi,
                tl = prediction.tl,
                score = prediction.score,
            )
        }
    }

    // region Pure helpers (exposed for testing)

    /**
     * Association window check. Strict `<` at 10s; negative deltas (clock
     * skew / wrapped) return false — a parity correction vs the pre-A5
     * code, which would have returned true for a future-relative state.
     */
    fun shouldRecordAssociation(
        state: NextWordPersistedState,
        nowMs: Long,
    ): Boolean {
        if (state.lastSelectedWord == null) return false
        val delta = nowMs - state.lastSelectionTimeMs
        return delta in 0 until ASSOCIATION_TIMEOUT_MS
    }

    /**
     * Split a compound word on `-` OR whitespace. Android preserves the
     * pre-A5 wrapper-side `split("-", " ")` semantics; iOS splits on `-`
     * only. Intentional divergence per `nextword-engine-boundary.md` §13.
     */
    fun splitCompound(word: String): List<String> {
        if (word.isEmpty()) return emptyList()
        return word.split("-", " ").filter { it.isNotEmpty() }
    }

    /**
     * Build sequential bigram pairs from a compound word. Order preserved
     * so the executor can feed them to `NextWordService.recordAssociation`
     * sequentially — parallel writes would race on the SQLite UNIQUE
     * constraint on `(prev_word, next_word, next_tl)`.
     */
    fun compoundAssociationPairs(
        displayText: String,
        roman: String,
    ): List<NextWordAssociationPair> {
        val parts = splitCompound(displayText)
        val romanParts = splitCompound(roman)
        if (parts.size <= 1) return emptyList()

        val pairs = ArrayList<NextWordAssociationPair>(parts.size - 1)
        for (i in 0 until parts.size - 1) {
            pairs.add(
                NextWordAssociationPair(
                    prev = parts[i],
                    prevTl = romanParts.getOrNull(i) ?: "",
                    next = parts[i + 1],
                    nextTl = romanParts.getOrNull(i + 1) ?: "",
                ),
            )
        }
        return pairs
    }

    /**
     * Punctuation / ASCII space / full-width space / pure-digit text never
     * triggers NextWord. Android superset per §13 addendum — scope limited
     * to the exact char set pre-A5 `NextWordHandler.NOISE_CHARS` contained:
     * [noisePunctuation] already bundles both space variants, so this
     * function does NOT fall back to `Char.isWhitespace()` (which would
     * widen the set to tab / newline / vertical tab — a behavior change
     * vs pre-A5).
     */
    fun isNoiseText(text: String): Boolean {
        if (text.isEmpty()) return true
        return text.all { it in noisePunctuation || (it.code in '0'.code..'9'.code) }
    }

    fun isSentenceEndPunctuation(text: String): Boolean {
        val firstChar = text.firstOrNull() ?: return false
        return firstChar in sentenceEndPunctuation
    }

    // endregion

    // region Intent handlers

    private fun decideWordSelected(
        text: String,
        roman: String,
        requireRomanMode: Boolean,
        triggerPrediction: Boolean,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ): NextWordOutcome {
        // Enter commits raw romanization only; skip entirely in Hanji mode.
        if (requireRomanMode && input.settings.isTranslateSwapped) {
            return NextWordOutcome(newState = state, effects = emptyList())
        }

        // Noise text (punctuation / whitespace / digits) never records or
        // predicts. Sentence-end is a subset of noise: branch on it first
        // so the reset path fires instead of the no-op path.
        if (text.isEmpty() || isNoiseText(text)) {
            return if (isSentenceEndPunctuation(text)) {
                resetAndClearPredictions(state)
            } else {
                NextWordOutcome(newState = state, effects = emptyList())
            }
        }

        // pojDisplayToTLDisplay is idempotent on TL input — safe for POJ and TPS alike.
        val textTl = TaigiPhonetics.pojDisplayToTLDisplay(roman)
        val prevTl = TaigiPhonetics.pojDisplayToTLDisplay(state.lastSelectedRoman ?: "")

        val effects = mutableListOf<NextWordOutcome.Effect>()

        if (input.settings.isAssociationRecordingEnabled) {
            val prevWord = state.lastSelectedWord
            if (prevWord != null && shouldRecordAssociation(state = state, nowMs = input.nowMs)) {
                effects.add(
                    NextWordOutcome.Effect.RecordAssociation(
                        NextWordAssociationPair(
                            prev = prevWord,
                            prevTl = prevTl,
                            next = text,
                            nextTl = textTl,
                        ),
                    ),
                )
            }
            val compound = compoundAssociationPairs(displayText = text, roman = textTl)
            if (compound.isNotEmpty()) {
                effects.add(NextWordOutcome.Effect.RecordCompoundAssociations(compound))
            }
        }

        effects.add(NextWordOutcome.Effect.RescheduleContextTimeout(afterMs = CONTEXT_TIMEOUT_MS))

        val newState =
            state.copy(
                lastSelectedWord = text,
                lastSelectedRoman = textTl,
                lastSelectionTimeMs = input.nowMs,
                currentGeneration = state.currentGeneration + 1,
            )

        if (triggerPrediction) {
            effects.add(
                NextWordOutcome.Effect.QueryPredictions(
                    word = text,
                    roman = textTl,
                    generation = newState.currentGeneration,
                    nowMs = input.nowMs,
                ),
            )
        }

        return NextWordOutcome(newState = newState, effects = effects)
    }

    private fun decideBackspace(
        lastChar: String,
        state: NextWordPersistedState,
        input: NextWordDecisionInput,
    ): NextWordOutcome {
        val newState =
            state.copy(
                lastSelectedWord = lastChar,
                lastSelectedRoman = null,
                lastSelectionTimeMs = input.nowMs,
                currentGeneration = state.currentGeneration + 1,
            )
        return NextWordOutcome(
            newState = newState,
            effects =
                listOf(
                    NextWordOutcome.Effect.QueryPredictions(
                        word = lastChar,
                        roman = "",
                        generation = newState.currentGeneration,
                        nowMs = input.nowMs,
                    ),
                ),
        )
    }

    private fun decideClearForNewComposing(state: NextWordPersistedState): NextWordOutcome {
        val newState =
            state.copy(
                isShowing = false,
                currentGeneration = state.currentGeneration + 1,
            )
        val effects: List<NextWordOutcome.Effect> =
            if (state.isShowing) {
                listOf(NextWordOutcome.Effect.ClearPredictionsUI(generation = newState.currentGeneration))
            } else {
                emptyList()
            }
        return NextWordOutcome(newState = newState, effects = effects)
    }

    /**
     * Shared reset path used by sentence-end punctuation, context timeout,
     * and `ResetFull` intents. Cancels the timeout, optionally clears the
     * UI, zeroes state, bumps generation.
     */
    private fun resetAndClearPredictions(state: NextWordPersistedState): NextWordOutcome {
        val newGeneration = state.currentGeneration + 1
        val newState = NextWordPersistedState.initial.copy(currentGeneration = newGeneration)

        val effects = mutableListOf<NextWordOutcome.Effect>(NextWordOutcome.Effect.CancelContextTimeout)
        if (state.isShowing) {
            effects.add(NextWordOutcome.Effect.ClearPredictionsUI(generation = newGeneration))
        }
        return NextWordOutcome(newState = newState, effects = effects)
    }

    // endregion
}
